### PR TITLE
Add workspace MIME handler and loading/saving workspaces manually

### PR DESCRIPTION
**Pull request by wonsjb**
## References

This pull request:

* adds a MIME handler for the `.jupyterlab-workspace` file extension that loads the workspace and redirects the user
* adds the ability to save the current workspace in the current working directory of the primary file browser
* adds some file organization and clean up fixes to the `apputils` and `apputils-extension` packages

cf. #6944 

## Code changes

1. Adds `Save Current Workspace` and `Save Current Workspace As...` items in the `File` menu
2. MIME handler for `.jupyterlab-workspace` files. When double clicked this loads the content of the workspace file and changes the URL to point to the opened workspace

## User-facing changes

workspaces

## Backwards-incompatible changes

None known

### CODE REVIEW

**Comment by afshin** [comment]
CC: @isabela-pf

File: packages/apputils-extension/src/workspace-ui.ts [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+/**
+ * This widget factory is used to handle double click on workspace
+ */
+class WorkspaceFactory extends ABCWidgetFactory<IDocumentWidget> {
+  workspaces: WorkspaceManager;
+  router: IRouter;
+  state: IStateDB;
+
+  constructor(workspaces: WorkspaceManager, router: IRouter, state: IStateDB) {
```
Comment (blink1073) [inline packages/apputils-extension/src/workspace-ui.ts:45]: Missing docstring

File: packages/apputils-extension/src/workspace-ui.ts [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+      fileTypes: [WORKSPACE_NAME],
+      defaultFor: [WORKSPACE_NAME],
+      readOnly: true
+    });
+    this.workspaces = workspaces;
+    this.router = router;
+    this.state = state;
+  }
+
+  protected createNewWidget(
```
Comment (blink1073) [inline packages/apputils-extension/src/workspace-ui.ts:57]: missing docstring

File: packages/apputils-extension/src/workspace-ui.ts [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+  } else {
+    return null;
+  }
+}
+
+/**
+ * A widget that gets a file path from a user.
+ */
+class SaveWidget extends Widget {
+  constructor(path: string) {
```
Comment (blink1073) [inline packages/apputils-extension/src/workspace-ui.ts:98]: Docstring

File: packages/apputils-extension/src/workspace-ui.ts [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    super({ node: createSaveNode(path) });
+  }
+
+  getValue(): string {
```
Comment (blink1073) [inline packages/apputils-extension/src/workspace-ui.ts:102]: Docstring

**Comment by isabela-pf** [comment]
I was going to say that I think workspaces fit better under Settings in the menu bar (because they will trigger a reload and shift in appearance like JupyterLab Themes), but since the only option in the menu bar is to save workspaces for now I think this could work under File. If there is interest in putting more workspace options in the menu bar though, I would strongly suggest putting it under Settings.

I do have a few questions about the interactions.

1. I really like just being able to click on the appropriate file in the file browser to open a workspace. Is there any indicator that that JSON file is a workspace besides the Jupyter logo as its' icon? 

2. Do the Save Workspace (and Save Workspace As) have any cases where they would need to be inactive (or similarly not be able to save)?

Other than that, I think this will work for the feature set listed here. Lovely work!

**Comment by wonsjb** [comment]
Hi Isabela,

For the workspace location, I've basically followed what I'm used to, (intellij, pycharm, visual studio code), where the workspace operations are under the file menu.

> I really like just being able to click on the appropriate file in the file browser to open a workspace. Is there any indicator that that JSON file is a workspace besides the Jupyter logo as its' icon?
The only other indication is the file extension name: .jupyterlab-workspace
Note: this is the same file file extension as jupyterlab's internal workspace location.

> Do the Save Workspace (and Save Workspace As) have any cases where they would need to be inactive (or similarly not be able to save)?
I did not see any. There is always a current workspace being there, so it should always save-able (as far as I know, at least)

Does that make sense ?

Hi Steven,

I've added the missing docstring, let me know what you think!

Thanks all for the review!

JB

**Comment by blink1073** [comment]
LGTM, I'll wait to see if @isabela-pf has any more concerns.

**Comment by isabela-pf** [comment]
Thanks for the explanation @wonsjb! I think that all makes sense, especially since is the beginning of features requested in #6944. I think it's good to merge.

**Pull request merged**